### ISSUE TITLE
fuzz test panic at "attempt to add/subtract with overflow"

### ISSUE THREAD

**Issue by you06** [tikv/tikv#6634]
## Bug Report

tikv panic while doing fuzz test, here is the stack backtrace

### attempt to add with overflow

details https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_continue_fuzz/detail/tikv_continue_fuzz/102/pipeline/36

```log
[2020-02-17T08:30:35.775Z] thread '<unnamed>' panicked at 'attempt to add with overflow', components/tidb_query/src/codec/mysql/duration.rs:212:30
[2020-02-17T08:30:35.775Z] stack backtrace:
[2020-02-17T08:30:35.775Z]    0: backtrace::backtrace::libunwind::trace
[2020-02-17T08:30:35.775Z]              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
[2020-02-17T08:30:35.775Z]    1: backtrace::backtrace::trace_unsynchronized
[2020-02-17T08:30:35.775Z]              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
[2020-02-17T08:30:35.775Z]    2: std::sys_common::backtrace::_print_fmt
[2020-02-17T08:30:35.775Z]              at src/libstd/sys_common/backtrace.rs:84
[2020-02-17T08:30:35.775Z]    3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
[2020-02-17T08:30:35.775Z]              at src/libstd/sys_common/backtrace.rs:61
[2020-02-17T08:30:35.775Z]    4: core::fmt::write
[2020-02-17T08:30:35.775Z]              at src/libcore/fmt/mod.rs:1025
[2020-02-17T08:30:35.776Z]    5: std::io::Write::write_fmt
[2020-02-17T08:30:35.776Z]              at src/libstd/io/mod.rs:1426
[2020-02-17T08:30:35.776Z]    6: std::sys_common::backtrace::_print
[2020-02-17T08:30:35.776Z]              at src/libstd/sys_common/backtrace.rs:65
[2020-02-17T08:30:35.776Z]    7: std::sys_common::backtrace::print
[2020-02-17T08:30:35.776Z]              at src/libstd/sys_common/backtrace.rs:50
[2020-02-17T08:30:35.776Z]    8: std::panicking::default_hook::{{closure}}
[2020-02-17T08:30:35.776Z]              at src/libstd/panicking.rs:193
[2020-02-17T08:30:35.776Z]    9: std::panicking::default_hook
[2020-02-17T08:30:35.776Z]              at src/libstd/panicking.rs:210
[2020-02-17T08:30:35.776Z]   10: std::panicking::rust_panic_with_hook
[2020-02-17T08:30:35.776Z]              at src/libstd/panicking.rs:471
[2020-02-17T08:30:35.776Z]   11: rust_begin_unwind
[2020-02-17T08:30:35.776Z]              at src/libstd/panicking.rs:375
[2020-02-17T08:30:35.776Z]   12: core::panicking::panic_fmt
[2020-02-17T08:30:35.776Z]              at src/libcore/panicking.rs:84
[2020-02-17T08:30:35.776Z]   13: core::panicking::panic
[2020-02-17T08:30:35.776Z]              at src/libcore/panicking.rs:51
[2020-02-17T08:30:35.776Z]   14: tidb_query::codec::mysql::duration::parser::parse
[2020-02-17T08:30:35.776Z]   15: tidb_query::codec::mysql::duration::Duration::parse
[2020-02-17T08:30:35.776Z]   16: fuzz_targets::fuzz_coprocessor_codec_duration_from_parse
[2020-02-17T08:30:35.776Z]   17: rust_fuzzer_test_input
[2020-02-17T08:30:35.776Z]   18: __rust_maybe_catch_panic
[2020-02-17T08:30:35.776Z]              at src/libpanic_unwind/lib.rs:78
[2020-02-17T08:30:35.776Z]   19: std::panicking::try
[2020-02-17T08:30:35.776Z]   20: LLVMFuzzerTestOneInput
[2020-02-17T08:30:35.776Z]   21: _ZN6fuzzer6Fuzzer15ExecuteCallbackEPKhm
[2020-02-17T08:30:35.776Z]              at libfuzzer/FuzzerLoop.cpp:515
[2020-02-17T08:30:35.776Z]   22: _ZN6fuzzer6Fuzzer6RunOneEPKhmbPNS_9InputInfoEPb
[2020-02-17T08:30:35.776Z]              at libfuzzer/FuzzerLoop.cpp:440
[2020-02-17T08:30:35.776Z]   23: _ZN6fuzzer6Fuzzer16MutateAndTestOneEv
[2020-02-17T08:30:35.776Z]              at libfuzzer/FuzzerLoop.cpp:649
[2020-02-17T08:30:35.776Z]   24: _ZN6fuzzer6Fuzzer4LoopERKSt6vectorISsNS_16fuzzer_allocatorISsEEE
[2020-02-17T08:30:35.776Z]              at libfuzzer/FuzzerLoop.cpp:775
[2020-02-17T08:30:35.776Z]   25: _ZN6fuzzer12FuzzerDriverEPiPPPcPFiPKhmE
[2020-02-17T08:30:35.776Z]   26: main
[2020-02-17T08:30:35.776Z]              at libfuzzer/FuzzerMain.cpp:20
[2020-02-17T08:30:35.776Z]   27: __libc_start_main
[2020-02-17T08:30:35.776Z]   28: <unknown>
[2020-02-17T08:30:35.776Z] note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2020-02-17T08:30:35.776Z] ==3236== ERROR: libFuzzer: deadly signal
[2020-02-17T08:30:35.776Z]     #0 0x55f9f0fd42b1  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x79a2b1)
[2020-02-17T08:30:35.776Z]     #1 0x55f9f3c2945f  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33ef45f)
[2020-02-17T08:30:35.776Z]     #2 0x55f9f3c2931f  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33ef31f)
[2020-02-17T08:30:35.776Z]     #3 0x55f9f3c3c817  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x3402817)
[2020-02-17T08:30:35.776Z]     #4 0x7f3bafa1c5cf  (/lib64/libpthread.so.0+0xf5cf)
[2020-02-17T08:30:35.776Z]     #5 0x7f3baf460206  (/lib64/libc.so.6+0x36206)
[2020-02-17T08:30:35.776Z]     #6 0x7f3baf4618f7  (/lib64/libc.so.6+0x378f7)
[2020-02-17T08:30:35.776Z]     #7 0x55f9f3c64716  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x342a716)
[2020-02-17T08:30:35.776Z]     #8 0x55f9f3c5de65  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x3423e65)
[2020-02-17T08:30:35.776Z]     #9 0x55f9f3c13bca  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33d9bca)
[2020-02-17T08:30:35.776Z]     #10 0x55f9f3c2a8ff  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33f08ff)
[2020-02-17T08:30:35.776Z]     #11 0x55f9f3c2a3c8  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33f03c8)
[2020-02-17T08:30:35.776Z]     #12 0x55f9f3c2b288  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33f1288)
[2020-02-17T08:30:35.776Z]     #13 0x55f9f3c2bdca  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33f1dca)
[2020-02-17T08:30:35.776Z]     #14 0x55f9f3c198b4  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33df8b4)
[2020-02-17T08:30:35.776Z]     #15 0x55f9f3c1376c  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x33d976c)
[2020-02-17T08:30:35.776Z]     #16 0x7f3baf44c3d4  (/lib64/libc.so.6+0x223d4)
[2020-02-17T08:30:35.776Z]     #17 0x55f9f0f6103b  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_parse+0x72703b)
[2020-02-17T08:30:35.776Z] 
[2020-02-17T08:30:35.776Z] NOTE: libFuzzer has rudimentary signal handlers.
[2020-02-17T08:30:35.777Z]       Combine libFuzzer with AddressSanitizer or similar for better crash reports.
[2020-02-17T08:30:35.777Z] SUMMARY: libFuzzer: deadly signal
[2020-02-17T08:30:35.777Z] MS: 1 CopyPart-; base unit: eb9dd91b4be9ae02234a121340c1551927342e31
[2020-02-17T08:30:35.777Z] 0x1,0x20,0x20,0x20,0x2d,0x20,0x20,0x34,0x32,0x34,0x34,0x32,0x34,0x32,0x20,0x20,0x34,0x32,0x34,0x34,0x32,0x34,0x34,0x34,0x34,0x34,0x5f,0x34,0x36,0x32,0x34,0x34,0x7e,
[2020-02-17T08:30:35.777Z] \x01   -  4244242  4244244444_46244~
[2020-02-17T08:30:35.777Z] artifact_prefix='./'; Test unit written to ./crash-cbcb433204a0d1b86adb370cf3a1e63920aa259b
[2020-02-17T08:30:35.777Z] Base64: ASAgIC0gIDQyNDQyNDIgIDQyNDQyNDQ0NDRfNDYyNDR+
[2020-02-17T08:30:36.034Z] Running fuzzer failed: Libfuzzer exited with code Some(77)
```

### attempt to subtract with overflow

details https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_continue_fuzz/detail/tikv_continue_fuzz/102/pipeline/35

```log
[2020-02-17T08:28:24.081Z] thread '<unnamed>' panicked at 'attempt to subtract with overflow', /rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libcore/num/mod.rs:2019:17
[2020-02-17T08:28:24.081Z] stack backtrace:
[2020-02-17T08:28:24.081Z]    0: backtrace::backtrace::libunwind::trace
[2020-02-17T08:28:24.081Z]              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
[2020-02-17T08:28:24.081Z]    1: backtrace::backtrace::trace_unsynchronized
[2020-02-17T08:28:24.081Z]              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
[2020-02-17T08:28:24.081Z]    2: std::sys_common::backtrace::_print_fmt
[2020-02-17T08:28:24.081Z]              at src/libstd/sys_common/backtrace.rs:84
[2020-02-17T08:28:24.081Z]    3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
[2020-02-17T08:28:24.081Z]              at src/libstd/sys_common/backtrace.rs:61
[2020-02-17T08:28:24.081Z]    4: core::fmt::write
[2020-02-17T08:28:24.081Z]              at src/libcore/fmt/mod.rs:1025
[2020-02-17T08:28:24.081Z]    5: std::io::Write::write_fmt
[2020-02-17T08:28:24.082Z]              at src/libstd/io/mod.rs:1426
[2020-02-17T08:28:24.082Z]    6: std::sys_common::backtrace::_print
[2020-02-17T08:28:24.082Z]              at src/libstd/sys_common/backtrace.rs:65
[2020-02-17T08:28:24.082Z]    7: std::sys_common::backtrace::print
[2020-02-17T08:28:24.082Z]              at src/libstd/sys_common/backtrace.rs:50
[2020-02-17T08:28:24.082Z]    8: std::panicking::default_hook::{{closure}}
[2020-02-17T08:28:24.082Z]              at src/libstd/panicking.rs:193
[2020-02-17T08:28:24.082Z]    9: std::panicking::default_hook
[2020-02-17T08:28:24.082Z]              at src/libstd/panicking.rs:210
[2020-02-17T08:28:24.082Z]   10: std::panicking::rust_panic_with_hook
[2020-02-17T08:28:24.082Z]              at src/libstd/panicking.rs:471
[2020-02-17T08:28:24.082Z]   11: rust_begin_unwind
[2020-02-17T08:28:24.082Z]              at src/libstd/panicking.rs:375
[2020-02-17T08:28:24.082Z]   12: core::panicking::panic_fmt
[2020-02-17T08:28:24.082Z]              at src/libcore/panicking.rs:84
[2020-02-17T08:28:24.082Z]   13: core::panicking::panic
[2020-02-17T08:28:24.082Z]              at src/libcore/panicking.rs:51
[2020-02-17T08:28:24.082Z]   14: tidb_query::codec::mysql::duration::Duration::from_nanos
[2020-02-17T08:28:24.082Z]   15: fuzz_targets::fuzz_coprocessor_codec_duration_from_nanos
[2020-02-17T08:28:24.082Z]   16: rust_fuzzer_test_input
[2020-02-17T08:28:24.082Z]   17: __rust_maybe_catch_panic
[2020-02-17T08:28:24.082Z]              at src/libpanic_unwind/lib.rs:78
[2020-02-17T08:28:24.082Z]   18: std::panicking::try
[2020-02-17T08:28:24.082Z]   19: LLVMFuzzerTestOneInput
[2020-02-17T08:28:24.082Z]   20: _ZN6fuzzer6Fuzzer15ExecuteCallbackEPKhm
[2020-02-17T08:28:24.082Z]              at libfuzzer/FuzzerLoop.cpp:515
[2020-02-17T08:28:24.082Z]   21: _ZN6fuzzer6Fuzzer6RunOneEPKhmbPNS_9InputInfoEPb
[2020-02-17T08:28:24.082Z]              at libfuzzer/FuzzerLoop.cpp:440
[2020-02-17T08:28:24.082Z]   22: _ZN6fuzzer6Fuzzer16MutateAndTestOneEv
[2020-02-17T08:28:24.082Z]              at libfuzzer/FuzzerLoop.cpp:649
[2020-02-17T08:28:24.082Z]   23: _ZN6fuzzer6Fuzzer4LoopERKSt6vectorISsNS_16fuzzer_allocatorISsEEE
[2020-02-17T08:28:24.082Z]              at libfuzzer/FuzzerLoop.cpp:775
[2020-02-17T08:28:24.082Z]   24: _ZN6fuzzer12FuzzerDriverEPiPPPcPFiPKhmE
[2020-02-17T08:28:24.082Z]   25: main
[2020-02-17T08:28:24.082Z]              at libfuzzer/FuzzerMain.cpp:20
[2020-02-17T08:28:24.082Z]   26: __libc_start_main
[2020-02-17T08:28:24.082Z]   27: <unknown>
[2020-02-17T08:28:24.082Z] note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2020-02-17T08:28:24.082Z] ==143307== ERROR: libFuzzer: deadly signal
[2020-02-17T08:28:24.082Z]     #0 0x561c818932b1  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x7992b1)
[2020-02-17T08:28:24.082Z]     #1 0x561c844c295b  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33c895b)
[2020-02-17T08:28:24.082Z]     #2 0x561c844c281b  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33c881b)
[2020-02-17T08:28:24.082Z]     #3 0x561c844d49e0  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33da9e0)
[2020-02-17T08:28:24.082Z]     #4 0x7f8baedee5cf  (/lib64/libpthread.so.0+0xf5cf)
[2020-02-17T08:28:24.082Z]     #5 0x7f8bae832206  (/lib64/libc.so.6+0x36206)
[2020-02-17T08:28:24.082Z]     #6 0x7f8bae8338f7  (/lib64/libc.so.6+0x378f7)
[2020-02-17T08:28:24.082Z]     #7 0x561c844fd7e6  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x34037e6)
[2020-02-17T08:28:24.082Z]     #8 0x561c844f6f35  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33fcf35)
[2020-02-17T08:28:24.082Z]     #9 0x561c844d525a  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33db25a)
[2020-02-17T08:28:24.082Z]     #10 0x561c844c3dfb  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33c9dfb)
[2020-02-17T08:28:24.082Z]     #11 0x561c844c38c4  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33c98c4)
[2020-02-17T08:28:24.082Z]     #12 0x561c844c4784  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33ca784)
[2020-02-17T08:28:24.082Z]     #13 0x561c844c52c6  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33cb2c6)
[2020-02-17T08:28:24.082Z]     #14 0x561c844b24a6  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33b84a6)
[2020-02-17T08:28:24.082Z]     #15 0x561c844ac83c  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x33b283c)
[2020-02-17T08:28:24.082Z]     #16 0x7f8bae81e3d4  (/lib64/libc.so.6+0x223d4)
[2020-02-17T08:28:24.082Z]     #17 0x561c8182003b  (/home/jenkins/agent/workspace/tikv_continue_fuzz/tikv/target/x86_64-unknown-linux-gnu/debug/fuzz_coprocessor_codec_duration_from_nanos+0x72603b)
[2020-02-17T08:28:24.082Z] 
[2020-02-17T08:28:24.082Z] NOTE: libFuzzer has rudimentary signal handlers.
[2020-02-17T08:28:24.082Z]       Combine libFuzzer with AddressSanitizer or similar for better crash reports.
[2020-02-17T08:28:24.082Z] SUMMARY: libFuzzer: deadly signal
[2020-02-17T08:28:24.082Z] MS: 5 CopyPart-CrossOver-ChangeByte-PersAutoDict-ChangeBit- DE: "\x00\x00\x00\x00\x00\x00\x00\x00"-; base unit: 455dff1cfd9a2e868cad5005e9313fd3c0b4914b
[2020-02-17T08:28:24.083Z] 0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x80,0x0,0x0,0x4,0x0,0xff,0xff,0x0,0x0,0x0,0xff,0x24,0xff,0xff,0xff,0x8f,
[2020-02-17T08:28:24.083Z] \x00\x00\x00\x00\x00\x00\x00\x80\x00\x00\x04\x00\xff\xff\x00\x00\x00\xff$\xff\xff\xff\x8f
[2020-02-17T08:28:24.083Z] artifact_prefix='./'; Test unit written to ./crash-a170c323594ea74498856eec07069468d2dd9131
[2020-02-17T08:28:24.083Z] Base64: AAAAAAAAAIAAAAQA//8AAAD/JP///48=
[2020-02-17T08:28:24.083Z] Running fuzzer failed: Libfuzzer exited with code Some(77)
```

### What version of TiKV are you using?

Code https://github.com/nrc/tikv/tree/fix-fuzz

### What operating system and CPU are you using?

* Operating system: CentOS Linux 7
* CPU: Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz

### What did you expect?

not panic

### What did happened?

panic

**Comment by iosmanthus** [comment]
It should be fixed by https://github.com/tikv/tikv/pull/6636

**Comment by andylokandy** [comment]
Closed by #6636
